src/iam/certhandler/modules/pkcs11.cpp

```diff
@@ -0,0 +1,997 @@
+    if (!err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
+    }
+
+    StaticString<uuid::cUUIDStrLen> uuid;
+
+    err = url.Search<1>("id=([^;&]*)", uuid);
+    if (!err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
+    }
+
+    Tie(id, err) = mUUIDManager->StringToUUID(uuid);
+    if (!err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
+    }
+
+    return Error::Enum::eNone;
+}
+
+Error PKCS11Module::GetValidInfo(pkcs11::SessionContext& session, Array<SearchObject>& certs,
+    Array<SearchObject>& privKeys, Array<SearchObject>& pubKeys, Array<CertInfo>& resCerts)
+{
+    for (auto privKey = privKeys.begin(); privKey != privKeys.end();) {
+        LOG_DBG() << "Private key found: ID = " << privKey->mID;
+
+        auto pubKey = FindObjectByID(pubKeys, privKey->mID);
+        if (pubKey == pubKeys.end()) {
+            privKey++;
+            continue;
+        }
+
+        LOG_DBG() << "Public key found: ID = " << privKey->mID;
+
+        auto cert = FindObjectByID(certs, privKey->mID);
+        if (cert == certs.end()) {
+            privKey++;
+            continue;
+        }
+
+        LOG_DBG() << "Certificate found: ID = " << privKey->mID;
+
+        // create certInfo
+        auto     x509Cert = MakeUnique<crypto::x509::Certificate>(&mTmpObjAllocator);
+        CertInfo validCert;
+
+        Error err = GetX509Cert(session, cert->mHandle, *x509Cert);
+        if (!err.IsNone()) {
+            LOG_ERR() << "Can't get x509 certificate: ID = " << cert->mID;
+            return AOS_ERROR_WRAP(err);
+        }
+
+        err = CreateCertInfo(*x509Cert, privKey->mID, cert->mID, validCert);
+        if (!err.IsNone()) {
+            return err;
+        }
+
+        // update containers
+        err = resCerts.PushBack(validCert);
+        if (!err.IsNone()) {
+            return AOS_ERROR_WRAP(err);
+        }
+
+        auto ret = certs.Remove(cert);
+        if (!ret.mError.IsNone()) {
+            return AOS_ERROR_WRAP(ret.mError);
+        }
+
+        ret = pubKeys.Remove(pubKey);
+        if (!ret.mError.IsNone()) {
+            return AOS_ERROR_WRAP(ret.mError);
+        }
+
+        ret = privKeys.Remove(privKey);
+        if (!ret.mError.IsNone()) {
+            return AOS_ERROR_WRAP(ret.mError);
+        }
+
+        privKey = ret.mValue;
+    }
+
+    return Error::Enum::eNone;
+}
+
+PKCS11Module::SearchObject* PKCS11Module::FindObjectByID(Array<SearchObject>& array, const Array<uint8_t>& id)
+{
+    SearchObject* result = array.begin();
+    while (result != array.end()) {
+        if (result->mID == id) {
+            return result;
+        }
+    }
+
+    return result;
+}
+
+Error PKCS11Module::GetX509Cert(
+    pkcs11::SessionContext& session, pkcs11::ObjectHandle object, crypto::x509::Certificate& cert)
+{
+    static constexpr auto cSingleAttribute = 1;
+
+    auto certBuffer = MakeUnique<DERCert>(&mTmpObjAllocator);
+
+    StaticArray<pkcs11::AttributeType, cSingleAttribute> types;
```

Comment:
Empty line after var definition and combine var block:
```c++
    StaticArray<pkcs11::AttributeType, cSingleAttribute> types;
    StaticArray<Array<uint8_t>, cSingleAttribute> values;

    types.PushBack(CKA_VALUE);
    values.PushBack(*certBuffer);
```

Reply:
fixed